src/modules/accounting/accounting.py

```diff
@@ -0,0 +1,210 @@
+
+        return report_data
+
+    @lru_cache(maxsize=1)
+    def _get_consensus_lido_state(self, blockstamp: BlockStamp) -> tuple[int, Gwei]:
+        lido_validators = self.w3.lido_validators.get_lido_validators(blockstamp)
+
+        count = len(lido_validators)
+        balance = Gwei(sum(int(validator.validator.balance) for validator in lido_validators))
+        return count, balance
+
+    @lru_cache(maxsize=1)
+    def _get_withdrawal_balance(self, blockstamp: BlockStamp) -> Wei:
+        return Wei(self.w3.eth.get_balance(
+            self.w3.lido_contracts.lido_locator.functions.withdrawalVault().call(
+                block_identifier=blockstamp.block_hash
+            ),
+            block_identifier=blockstamp.block_hash,
+        ))
+
+    @lru_cache(maxsize=1)
+    def _get_el_vault_balance(self, blockstamp: BlockStamp) -> Wei:
+        return Wei(self.w3.eth.get_balance(
+            self.w3.lido_contracts.lido_locator.functions.elRewardsVault().call(
+                block_identifier=blockstamp.block_hash
+            ),
+            block_identifier=blockstamp.block_hash,
+        ))
+
+    def _get_last_withdrawal_request_to_finalize(self, blockstamp: BlockStamp) -> int:
+        chain_config = self._get_chain_config(blockstamp)
+        frame_config = self._get_frame_config(blockstamp)
+
+        is_bunker = self._is_bunker(blockstamp)
+        withdrawal_vault_balance = self._get_withdrawal_balance(blockstamp)
+        el_rewards_vault_balance = self._get_el_vault_balance(blockstamp)
+        finalization_share_rate = self._get_finalization_shares_rate(blockstamp)
+
+        withdrawal_service = Withdrawal(self.w3, blockstamp, chain_config, frame_config)
+
+        return withdrawal_service.get_next_last_finalizable_id(
+            is_bunker,
+            finalization_share_rate,
+            withdrawal_vault_balance,
+            el_rewards_vault_balance,
+        )
+
+    @lru_cache(maxsize=1)
+    def _get_finalization_shares_rate(self, blockstamp: BlockStamp) -> int:
+        simulation = self.get_rebase_after_report(blockstamp)
+        return simulation.post_total_pooled_ether * SHARE_RATE_PRECISION_E27 // simulation.post_total_shares
+
+    @lru_cache(maxsize=1)
+    def get_rebase_after_report(self, blockstamp: BlockStamp) -> LidoReportRebase:
+        chain_conf = self._get_chain_config(blockstamp)
+        frame_config = self._get_frame_config(blockstamp)
+
+        last_ref_slot = self.report_contract.functions.getLastProcessingRefSlot().call(
+            block_identifier=blockstamp.block_hash,
+        )
+
+        if not last_ref_slot:
+            slots_elapsed = blockstamp.ref_slot - frame_config.initial_epoch * chain_conf.slots_per_epoch
+        else:
+            slots_elapsed = blockstamp.ref_slot - last_ref_slot
+
+        validators_count, cl_balance = self._get_consensus_lido_state(blockstamp)
+
+        timestamp = chain_conf.genesis_time + blockstamp.ref_slot * chain_conf.seconds_per_slot
```

Comment:
consider having a util function set for these operations, like:
get_timestamp_by_slot(slot, chain_conf)
get_first_slot_in_epoch
get_epoch_by_slot
get_frame_by_slot
and so on